### PR TITLE
detect dev for install command

### LINKED ISSUE #1412: Respect 'dev' keyword in packagist UI to show composer require --dev
Composer itself recognizes [few keywords to distinguish dev packages](https://getcomposer.org/doc/04-schema.md#keywords) (like `dev` or `static analysis`), but packagist does not seem to respect those. I believe it should show `composer require --dev` for those. [Example package](https://packagist.org/packages/shipmonk/phpstan-rules).

image

### PR DESCRIPTION
**Pull request by chr-hertel**
Fixes #1412

### CODE REVIEW

File: src/Entity/Package.php [hunk 1]
```diff
@@ -687,6 +687,23 @@ public function getType(): string|null
...
 
+    public function getInstallCommand(): string
+    {
+        $command = 'create-project';
+
+        if ('project' !== $this->getType()) {
+            $command = 'require';
+
+            /** @var Version $version */
+            $version = $this->getVersions()->first();
```
Comment (chr-hertel) [inline src/Entity/Package.php:698]: not sure which version to use ...

File: src/Entity/Version.php [hunk 2]
```diff
@@ -635,6 +635,13 @@ public function getTags(): Collection
         return $this->tags;
     }
 
+    public function hasDevTag(): bool
+    {
+        return [] !== array_filter($this->getTags()->toArray(), static function (Tag $tag) {
```
Comment (stof) [inline src/Entity/Version.php:640]: looping over the tags and stopping when we find a dev tag would be more efficient than filtering the whole list all the time.

File: src/Entity/Package.php (hunk 1 above)
Comment (stof) [reply src/Entity/Package.php:698]: The `view_package` template already receive a `version` variable (which can be `null`). this is the version that should be used for the install command IMO (which would be done by keeping the command in the template as it is done right now, but adding a `{% if version and version.hasDevTag %} --dev{% endif %}`)

File: src/Entity/Version.php (hunk 2 above)
Comment (chr-hertel) [reply src/Entity/Version.php:640]: true

File: src/Entity/Package.php (hunk 1 above)
Comment (chr-hertel) [reply src/Entity/Package.php:698]: yea, i know that moving it from the template to PHP feels like the more complex solution at first, but i tend to kill every if in twig that i can replace with sth that is easily unit tested.

but sure, that's a debatable design decision

Comment (stof) [reply src/Entity/Package.php:698]: Twig already has the right version to use passed to it by the controller. The entity does not have it. So moving this to the entity itself looks like the bad decision IMO.

Comment (chr-hertel) [reply src/Entity/Package.php:698]: wasn't aware of that controller code to sort out the most relevant version - thanks for that pointer!
will give it another thought later :+1:

Comment (Seldaek) [reply src/Entity/Package.php:698]: IMO this function should just accept a version as argument, and then the template can pass it in.

**Comment by Seldaek** [comment]
Thanks!

File: templates/package/view_package.html.twig [hunk 3]
```diff
@@ -46,7 +46,7 @@
 
             <div class="row">
                 <div class="col-md-8">
-                    <p class="requireme"><i class="glyphicon glyphicon-save"></i> <input type="text" readonly="readonly" value="composer {% if package.type == 'project' %}create-project{% else %}require{% endif %} {{ "#{package.vendor}/#{package.packageName}" }}" /></p>
+                    <p class="requireme"><i class="glyphicon glyphicon-save"></i> <input type="text" readonly="readonly" value="{{ package.installCommand(expandedVersion) }}" /></p>
```
[the remark this answers is not in the document]
Comment (Seldaek) [reply templates/package/view_package.html.twig:49]: Yes true it should be version, I'll fix that.

**Pull request merged**